metis/lib/models/data_block.rb

```diff
@@ -88,22 +86,38 @@ def compute_hash!
             data_block_id: existing_block.id,
           )
 
+          yield [:temp_delete] if block_given?
           # destroy the redundant file
-          ::File.delete(location)
+          ::File.delete(location) if has_data?
 
           # destroy this redundant record
+          yield [:temp_destroy] if block_given?
           destroy
           return
+        else
+          # Create a stub in position without removing the exist file, to ensure that
+          # if the database transaction fails, we still have a way of finding the file backing
+          # this datablock.
+          yield [:sym, location, file_location(md5_hash)] if block_given?
+          begin
+            ::File.symlink(location, file_location(md5_hash))
+          rescue ::Errno::EEXIST => e
+            # Some other process may have already prepared the file by the same hash,
+          end
+
+          yield [:new_update] if block_given?
+          update(md5_hash: md5_hash)
         end
+      end
 
-        # Actually move the file
-        ::File.rename(
-          location,
-          file_location(md5_hash)
-        )
+      if ::File.symlink?(location) && !temp_hash?
+        original_file = ::File.readlink(location)
+        ::File.rename(original_file, location)
+      end
 
-        # Now update the record
-        update(md5_hash: md5_hash)
+      unless has_data?
```

Comment:
Is the effect of this that any 0-Byte data blocks will be automatically removed when the Metis command tries to calculate the md5 hash?

Reply:
has_data? just checks file existence so 0 byte files should be ok.